diff --git a/variants/vueapp/src/components/FilterResultsTable.vue b/variants/vueapp/src/components/FilterResultsTable.vue
index b3fb41de..84185be0 100644
--- a/variants/vueapp/src/components/FilterResultsTable.vue
+++ b/variants/vueapp/src/components/FilterResultsTable.vue
@@ -224,6 +224,42 @@ const tableRows = ref([])
 /** Whether the Vue3EasyDataTable is loading. */
 const tableLoading = ref(false)
 
+const getAcmgRating = (payload) => {
+  return acmgRatingStore.getAcmgRating(
+    new SeqvarImpl(
+      payload.release === 'GRCh37' ? 'grch37' : 'grch38',
+      payload.chromosome,
+      payload.start,
+      payload.reference,
+      payload.alternative,
+    ),
+  )
+}
+
+const getFlags = (payload) => {
+  return flagsStore.getFlags(
+    new SeqvarImpl(
+      payload.release === 'GRCh37' ? 'grch37' : 'grch38',
+      payload.chromosome,
+      payload.start,
+      payload.reference,
+      payload.alternative,
+    ),
+  )
+}
+
+const hasComments = (payload) => {
+  return commentsStore.hasComments(
+    new SeqvarImpl(
+      payload.release === 'GRCh37' ? 'grch37' : 'grch38',
+      payload.chromosome,
+      payload.start,
+      payload.reference,
+      payload.alternative,
+    ),
+  )
+}
+
 /**
  * Configuration for the row to color them based on flags.
  */
@@ -232,15 +268,7 @@ const tableRowClassName = (item, _rowNumber) => {
     return ''
   }
   const flagColors = ['positive', 'uncertain', 'negative']
-  const flags = flagsStore.getFlags(
-    new SeqvarImpl(
-      item.payload.genome_build == 'GRCh37' ? 'grch37' : 'grch38',
-      item.payload.chromosome,
-      item.payload.start,
-      item.payload.reference,
-      item.payload.alternative,
-    ),
-  )
+  const flags = getFlags(item.payload)
   if (!flags) {
     return ''
   }
@@ -591,7 +619,7 @@ watch(
               @click="showVariantDetails(sodar_uuid)"
             />
             <i-fa-solid-bookmark
-              v-if="flagsStore.getFlags(payload)"
+              v-if="getFlags(payload)"
               class="text-muted ml-1"
               title="flags & bookmarks"
               role="button"
@@ -606,7 +634,7 @@ watch(
             />
 
             <i-fa-solid-comment
-              v-if="commentsStore.hasComments(payload)"
+              v-if="hasComments(payload)"
               class="text-muted ml-1"
               role="button"
               @click="showVariantDetails(sodar_uuid, 'comments')"
@@ -621,11 +649,11 @@ watch(
             <span
               title="ACMG rating"
               :class="
-                getAcmgBadgeClasses(acmgRatingStore.getAcmgRating(payload))
+                getAcmgBadgeClasses(getAcmgRating(payload))
               "
               role="button"
               @click="showVariantDetails(sodar_uuid, 'acmg-rating')"
-              >{{ acmgRatingStore.getAcmgRating(payload) || '-' }}</span
+              >{{ getAcmgRating(payload) || '-' }}</span
             >
 
             <a
