{% load variants_tags %}

{# Variant filter table header #}
<tr>
  <th class="p-0 m-0"></th> {# fold-out #}
  <th class="p-0 m-0"></th> {# rank #}
  <th class="p-0 m-0"></th> {# checkbox #}
  <th class="p-0 m-0"></th> {# bookmark #}
  <th class="p-0 m-0"></th> {# databases #}
  <th colspan="3" class="detail-info-coordinates">Coordinates</th>
  <th colspan="3" class="detail-info-clinvar">ClinVar</th>
  <th colspan="2" class="detail-frequencies-exac">ExAC</th>
  <th colspan="2" class="detail-frequencies-gnomad-exomes">gnomAD&nbsp;exomes</th>
  <th colspan="2" class="detail-frequencies-gnomad-genomes">gnomAD&nbsp;genomes</th>
  <th colspan="2" class="detail-frequencies-thousand-genomes">1000&nbsp;genomes</th>
  <th colspan="2" class="detail-frequencies-inhouse">in-house</th>
  <th colspan="2" class="detail-frequencies-mtdb">mtDB</th>
  <th colspan="2" class="detail-frequencies-helixmtdb">HelixMTdb</th>
  <th colspan="2" class="detail-frequencies-mitomap">MITOMAP</th>
  <th class="detail-constraints-exac-pli">ExAC</th>
  <th class="detail-constraints-exac-z-mis">ExAC</th>
  <th class="detail-constraints-exac-z-syn">ExAC</th>
  <th class="detail-constraints-gnomad-loeuf">gnomAD</th>
  <th class="detail-constraints-gnomad-pli">gnomAD</th>
  <th class="detail-constraints-gnomad-z-mis">gnomAD</th>
  <th class="detail-constraints-gnomad-z-syn">gnomAD</th>
  <th></th>{# gene #}
  <th></th>{# disease gene #}
  <th class="effect-change"></th>{# effect change #}
  <th class="effect-text"></th>{# effect text #}
  <th class="effect-change-protein"></th>{# effect change protein #}
  <th class="effect-change-cdna"></th>{# effect change cdna #}
  <th class="distance-to-splicesite"></th>{# distance to splicesite #}
  {% for value in result_extra_annos_header %}
    <th class="extra_annos-{{ forloop.counter }}"></th>
  {% endfor %}
  {% if query_type == "project" %}
    <th></th> {# cases per gene #}
  {% endif %}
  {% if has_phenotype_scores %}
    <th colspan="2">phenotype</th>
  {% endif %}
  {% if has_pathogenicity_scores %}
    <th colspan="2">pathogenicity</th>
  {% endif %}
  {% if has_phenotype_scores and has_pathogenicity_scores %}
    <th colspan="2">pheno.&nbsp;&amp;&nbsp;patho.</th>
  {% endif %}
  {% if query_type == "case" %}
    {# pedigree members #}
    {% for member in pedigree %}
      <th></th>
    {% endfor %}
  {% else %}  {# query_type == "project" #}
    <th></th>
    <th></th>
    <th></th>
  {% endif %}
  <th></th> {# link-outs #}
</tr>
<tr>
  <th class="p-0 m-0"></th> {# fold-out #}
  <th class="p-0 m-0"></th> {# rank #}
  <th class="p-0 m-0"></th> {# checkbox #}
  <th class="p-0 m-0"></th> {# bookmark #}
  <th class="p-0 m-0"></th> {# databases #}
  <th class="detail-info-coordinates">position</th>
  <th class="detail-info-coordinates">ref</th>
  <th class="detail-info-coordinates">alt</th>
  <th class="detail-info-clinvar">sign.&nbsp;&amp;&nbsp;rating</th>
  <th class="detail-info-clinvar"></th>
  <th class="detail-info-clinvar p-0"></th>
  <th class="detail-frequencies-exac">frequency</th>
  <th class="detail-frequencies-gnomad-exomes">frequency</th>
  <th class="detail-frequencies-gnomad-genomes">frequency</th>
  <th class="detail-frequencies-thousand-genomes">frequency</th>
  <th class="detail-frequencies-exac">#hom.</th>
  <th class="detail-frequencies-gnomad-exomes">#hom.</th>
  <th class="detail-frequencies-gnomad-genomes">#hom.</th>
  <th class="detail-frequencies-thousand-genomes">#hom.</th>
  <th class="detail-frequencies-inhouse">#carriers</th>
  <th class="detail-frequencies-inhouse">#hom.</th>
  <th class="detail-frequencies-mtdb">frequency</th>
  <th class="detail-frequencies-mtdb">#carriers.</th>
  <th class="detail-frequencies-helixmtdb">frequency</th>
  <th class="detail-frequencies-helixmtdb">#hom.</th>
  <th class="detail-frequencies-mitomap">frequency</th>
  <th class="detail-frequencies-mitomap">#carriers</th>
  <th class="detail-constraints-exac-pli">pLI</th>
  <th class="detail-constraints-exac-z-mis">Z mis</th>
  <th class="detail-constraints-exac-z-syn">Z syn</th>
  <th class="detail-constraints-gnomad-loeuf">LOEUF</th>
  <th class="detail-constraints-gnomad-pli">pLI</th>
  <th class="detail-constraints-gnomad-z-mis">Z mis</th>
  <th class="detail-constraints-gnomad-z-syn">Z syn</th>
  <th>gene</th>
  <th data-toggle="tooltip" data-placement="top" data-html="true" title="<strong><u>D</u></strong>isease <strong><u>G</u></strong>ene; variant found in gene listed in OMIM.">DG</th>
  <th class="effect-change">effect</th>
  <th class="effect-text">effect text</th>
  <th class="effect-change-protein">effect protein</th>
  <th class="effect-change-cdna">effect cdna</th>
  <th class="distance-to-splicesite">dist. to splicesite</th>
  {% for value in result_extra_annos_header %}
    <th class="extra_annos-{{ forloop.counter }}"> {{ value }}</th>
  {% endfor %}
  {% if query_type == "project" %}
    <th data-toggle="tooltip" data-placement="top" data-html="true" title="Number affected cases per gene.">
      <sup>#c</sup>&frasl;<sub>g</sub>
    </th>
  {% endif %}
  {% if has_phenotype_scores %}
    <th>score</th>
    <th>rank</th>
  {% endif %}
  {% if has_pathogenicity_scores %}
    <th>score</th>
    <th>rank</th>
  {% endif %}
  {% if has_phenotype_scores and has_pathogenicity_scores %}
    <th>score</th>
    <th>rank</th>
  {% endif %}
  {% if query_type == "case" %}
    {% for member in pedigree %}
      <th style="max-width: 80px;">
        <div class="sodar-overflow-container sodar-overflow-hover">
          {{ member.patient|only_source_name }}
        </div>
      </th>
    {% endfor %}
  {% else %}  {# query_type == "project #}
    <th>family</th>
    <th>sample</th>
    <th>genotype</th>
  {% endif %}
  <th></th> {# link-outs #}
</tr>
