{# Variant filter table row #}

{% load humanize %}
{% load dict %}
{% load variants_tags %}
{% load projectroles_common_tags %}

{% get_app_setting "variants" "umd_predictor_api_token" user=user as umd_predictor_api_token %}
{% get_django_setting "PROJECTROLES_KIOSK_MODE" as kiosk_mode %}
{% get_django_setting "VARFISH_ENABLE_HGMD_PRO_LINKOUT" as hgmd_pro_enabled %}
{% get_django_setting "VARFISH_HGMD_PRO_LINKOUT_URL_PREFIX" as hgmd_pro_prefix %}
{% get_app_setting "variants" "ga4gh_beacon_network_widget_enabled" user=user as ga4gh_beacon_network_widget_enabled %}
{% ambiguous_frequency_warning entry exac_enabled thousand_genomes_enabled gnomad_exomes_enabled gnomad_genomes_enabled inhouse_enabled as warning %}

{% with symbol=entry|get_symbol %}
<tr class="variant-row {% if not training_mode %}variant-row-{{ entry|flag_class }}{% endif %}" id="{{ entry.case_uuid }}-{{ entry|smallvar_description }}">
  {# fold-out #}
  <td style="width: 0px" class="clickable toggle-variant-details pl-0 pr-0"
      data-url="{% if cohort %}
          {% url 'variants:small-variant-details-cohort' project=project.sodar_uuid cohort=cohort.sodar_uuid case=entry.case_uuid release=entry.release chromosome=entry.chromosome start=entry.start end=entry.end reference=entry.reference alternative=entry.alternative database=database gene_id=entry.gene_id ensembl_transcript_id=entry.ensembl_transcript_id training_mode=training_mode %}
        {% else %}
          {% url 'variants:small-variant-details' project=project.sodar_uuid case=entry.case_uuid release=entry.release chromosome=entry.chromosome start=entry.start end=entry.end reference=entry.reference alternative=entry.alternative database=database gene_id=entry.gene_id ensembl_transcript_id=entry.ensembl_transcript_id training_mode=training_mode %}
        {% endif %}">
    <img src="/icons/fa-solid/chevron-right.svg" />
  </td>
  {# rank #}
  <td style="width: 0px" class="text-nowrap text-right text-muted" data-order="{{ forloop.counter }}">
      #{{ forloop.counter }}
  </td>
  {# checkbox #}
  <td class="pl-0 pr-1">
    <input type="checkbox" value="{{ entry|smallvar_description }}" class="multivar-selector" data-case="{{ entry.case_uuid }}">
  </td>
  {# bookmark #}
  {% if not training_mode %}
  <td style="width: 0px" class="bookmark pl-0 pr-0 text-nowrap">
    <a style="text-decoration: none" class="variant-bookmark-comment-group" data-variant="{{ entry|smallvar_description }}" data-project="{{ project.sodar_uuid }}" data-case="{{ entry.case_uuid }}">
      <img class="variant-bookmark img-gray" src="/icons/fa-{% if entry.flag_count %}solid{% else %}regular{% endif %}/bookmark.svg" />
      <img class="variant-comment img-gray" src="/icons/fa-{% if entry.comment_count %}solid{% else %}regular{% endif %}/comment.svg" />
    </a>
    <span class="{{ entry|acmg_badge_class }} variant-acmg" style="width: 22px; display: inline-block"
          data-variant="{{ entry|smallvar_description }}" data-project="{{ project.sodar_uuid }}" data-case="{{ entry.case_uuid }}">
      {{ entry|acmg_classification|default:"-" }}
    </span>
  </td>
  {% else %}
  <td></td>
  {% endif %}
  {# databases #}
  <td style="width: 0px" class="pl-2 pr-0 text-nowrap">
    {# dbSNP Indicator #}
    {% if entry.rsid %}
      <a target="_blank" href="https://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs={{ entry.rsid|slice:'2:' }}">
    {% endif %}
      <i class="iconify text-muted" data-icon="fa-solid:database"
        {% if entry.rsid %}
          data-html="true"
          data-tooltip="tooltip"
          data-placement="top"
          title="See {{ entry.rsid }} in dbSNP"
       {% else %}
        style="opacity: 0.2"
       {% endif %}
      ></i>{% if entry.rsid %}</a>
    {% endif %}

    {% include "variants/filter_result/clinvar_indicator.html" %}
    {% include "variants/filter_result/hgmd_public_indicator.html" %}
  </td>
  <td style="width: 120px;" class="pl-2 detail-info-coordinates" data-order="{{ entry.chromosome_no|stringformat:"02d" }}-{{ entry.start|stringformat:"09d" }}">
    <div class="text-nowrap">
      {% entry_chr entry %}:{{ entry.start|intcomma }}
      {% if entry.chromosome_no == 25 %}
        {% if entry.start|check_mt_position_homopolymer or entry.end|check_mt_position_homopolymer %}
          <i class="iconify text-muted" data-icon="bi:exclamation-circle" data-toggle="tooltip" title='Variant in homopolymeric region'></i>
        {% endif %}
      {% endif %}
    </div>
  </td>
  <td style="max-width: 50px;" class="detail-info-coordinates">
    <div class="sodar-overflow-container sodar-overflow-hover">
      {{ entry.reference }}
    </div>
  </td>
  <td style="max-width: 50px;" class="detail-info-coordinates">
    <div class="sodar-overflow-container sodar-overflow-hover">
      {{ entry.alternative }}
    </div>
  </td>
  <td style="width: 50px;" class="pl-2 detail-info-clinvar" data-order="{% if entry.max_significance %}{{ entry.max_significance|significance_importance }}-{{ entry.max_clinvar_status|status_importance }}{% else %}9-9{% endif %}">
    {% if entry.pathogenicity %}
        <span class="badge-group">
          <span class="badge badge-{{ entry.pathogenicity|significance_color }}">
            {{ entry.pathogenicity }}
          </span>
          <span class="badge badge-dark" data-toggle="tooltip" data-placement="top" title="{{ entry.max_clinvar_status }}">
            {% for i in "1234" %}
              {% if forloop.counter <= entry.point_rating %}
                <i class="iconify" data-icon="fa-solid:star"></i>
              {% else %}
                <i class="iconify" data-icon="fa-regular:star"></i>
              {% endif %}
            {% endfor %}
          </span>
        </span>
    {% else %}
        -
    {% endif %}
  </td>
  <td style="max-width: 80px;" class="pl-2 detail-info-clinvar">
  </td>
  <td class="detail-info-clinvar p-0"></td>
  <td class="detail-frequencies-exac">
    {% if entry.chromosome_no == 25 %}
      -
    {% else %}
      {{ entry.exac_frequency|floatformat:5|default:"-" }}
      {% if warning %}
        <i class="iconify text-muted" data-icon="bi:exclamation-circle" data-toggle="tooltip" title='Table(s) {{ warning|join:", " }} contain(s) freq > 0.1 or #hom > 50'></i>
      {% endif %}
    {% endif %}
  </td>
  <td class="detail-frequencies-gnomad-exomes">
    {% if entry.chromosome_no == 25 %}
      -
    {% else %}
      {{ entry.gnomad_exomes_frequency|floatformat:5|default:"-" }}
      {% if warning %}
        <i class="text-muted iconify" data-icon="bi:exclamation-circle" data-toggle="tooltip" title='Table(s) {{ warning|join:", " }} contain(s) freq > 0.1 or #hom > 50'></i>
      {% endif %}
    {% endif %}
  </td>
  <td class="detail-frequencies-gnomad-genomes">
    {% if entry.chromosome_no == 25 %}
      -
    {% else %}
      {{ entry.gnomad_genomes_frequency|floatformat:5|default:"-" }}
      {% if warning %}
        <i class="iconify text-muted" data-icon="bi:exclamation-circle" data-toggle="tooltip" title='Table(s) {{ warning|join:", " }} contain(s) freq > 0.1 or #hom > 50'></i>
      {% endif %}
    {% endif %}
  </td>
  <td class="detail-frequencies-thousand-genomes">
    {% if entry.chromosome_no == 25 %}
      -
    {% else %}
      {{ entry.thousand_genomes_frequency|floatformat:5|default:"-" }}
      {% if warning %}
        <i class="iconify text-muted" data-icon="bi:exclamation-circle" data-toggle="tooltip" title='Table(s) {{ warning|join:", " }} contain(s) freq > 0.1 or #hom > 50'></i>
      {% endif %}
    {% endif %}
  </td>
  <td class="detail-frequencies-exac text-right">
    {% if entry.chromosome_no == 25 %}
      -
    {% else %}
      {{ entry.exac_homozygous|default_if_none:"-" }}
    {% endif %}
  </td>
  <td class="detail-frequencies-gnomad-exomes text-right">
    {% if entry.chromosome_no == 25 %}
      -
    {% else %}
      {{ entry.gnomad_exomes_homozygous|default_if_none:"-" }}
    {% endif %}
  </td>
  <td class="detail-frequencies-gnomad-genomes text-right">
    {% if entry.chromosome_no == 25 %}
      -
    {% else %}
      {{ entry.gnomad_genomes_homozygous|default_if_none:"-" }}
    {% endif %}
  </td>
  <td class="detail-frequencies-thousand-genomes text-right">
    {% if entry.chromosome_no == 25 %}
      -
    {% else %}
      {{ entry.thousand_genomes_homozygous|default_if_none:"-" }}
    {% endif %}
  </td>
  <td class="detail-frequencies-inhouse text-right">
    {{ entry.inhouse_carriers|default_if_none:"-" }}
    {% if warning %}
      <i class="iconify text-muted" data-icon="bi:exclamation-circle" data-toggle="tooltip" title='Table(s) {{ warning|join:", " }} contain(s) freq > 0.1 or #hom > 50'></i>
    {% endif %}
  </td>
  <td class="detail-frequencies-inhouse text-right">
    {{ entry.inhouse_hom_alt|default_if_none:"-" }}
  </td>
  <td class="detail-frequencies-mtdb text-right">
    {% if entry.chromosome_no == 25 %}
      {% if entry.mtdb_dloop %}
        <i class="iconify text-muted" data-icon="bi:exclamation-circle" data-toggle="tooltip" title='Variant is in D-loop region'></i>
      {% endif %}
      {{ entry.mtdb_frequency|floatformat:5|default_if_none:"-" }}
    {% else %}
      -
    {% endif %}
  </td>
  <td class="detail-frequencies-mtdb text-right">
    {% if entry.chromosome_no == 25 %}
      {{ entry.mtdb_count|default_if_none:"-" }}
    {% else %}
      -
    {% endif %}
  </td>
  <td class="detail-frequencies-helixmtdb text-right text-nowrap">
    {% if entry.chromosome_no == 25 %}
      {% if entry.helixmtdb_is_triallelic %}
        <i class="iconify text-muted" data-icon="bi:exclamation-circle" data-toggle="tooltip" title='Variant is part of a triallelic site'></i>
      {% endif %}
      {{ entry.helixmtdb_frequency|floatformat:5|default_if_none:"-" }}
    {% else %}
      -
    {% endif %}
  </td>
  <td class="detail-frequencies-helixmtdb text-right">
    {% if entry.chromosome_no == 25 %}
      {{ entry.helixmtdb_hom_count|default_if_none:"-" }}
    {% else %}
      -
    {% endif %}
  </td>
  <td class="detail-frequencies-mitomap text-right">
    {% if entry.chromosome_no == 25 %}
      {{ entry.mitomap_frequency|floatformat:5|default_if_none:"-" }}
    {% else %}
      -
    {% endif %}
  </td>
  <td class="detail-frequencies-mitomap text-right">
    {% if entry.chromosome_no == 25 %}
      {{ entry.mitomap_count|default_if_none:"-" }}
    {% else %}
      -
    {% endif %}
  </td>
  <td class="detail-constraints-exac-pli text-right">
    {{ entry.exac_pLI|floatformat:3|default:"-" }}
  </td>
  <td class="detail-constraints-exac-z-mis text-right">
    {{ entry.exac_mis_z|floatformat:3|default:"-" }}
  </td>
  <td class="detail-constraints-exac-z-syn text-right">
    {{ entry.exac_syn_z|floatformat:3|default:"-" }}
  </td>
  <td class="detail-constraints-gnomad-loeuf text-right" data-order="{{ entry.gnomad_loeuf|default:0 }}">
    <span
      {% if entry.gnomad_loeuf %}
        data-toggle="tooltip" data-html="true" title="<sup>o</sup>&frasl;<sub>e</sub> = {{ entry.gnomad_oe_lof }} <small>({{ entry.gnomad_oe_lof_lower }} - {{ entry.gnomad_oe_lof_upper }})</small>"
      {% endif %}>
      {{ entry.gnomad_loeuf|floatformat:3|default:"-" }}
    </span>
  </td>
  <td class="detail-constraints-gnomad-pli text-right">
    {{ entry.gnomad_pLI|floatformat:3|default:"-" }}
  </td>
  <td class="detail-constraints-gnomad-z-mis text-right">
    {{ entry.gnomad_mis_z|floatformat:3|default:"-" }}
  </td>
  <td class="detail-constraints-gnomad-z-syn text-right">
    {{ entry.gnomad_syn_z|floatformat:3|default:"-" }}
  </td>
  <td class="pl-3 pr-3 text-nowrap" data-order="{% if symbol %}{% if entry.acmg_symbol %}0{% else %}1{% endif %}-{{ symbol }}{% else %}~{% endif %}">
    {% if not symbol %}
      -
    {% else %}
      <div class="btn-group sodar-list-btn-group">
        <button
          type="button"
          title="Go to locus in IGV"
          class="btn btn-outline-dark sodar-list-btn">
          {{ symbol|default:"." }}
        </button>
        <button type="button" class="btn btn-sm btn-outline-dark dropdown-toggle dropdown-toggle-split" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
          <span class="sr-only">Toggle Dropdown</span>
        </button>
        <div class="dropdown-menu dropdown-menu-right" style="z-index: 1031;">
          <a class="dropdown-item" href="https://www.omim.org/search/?search={{ symbol }}" target="_blank">
            Gene @OMIM
          </a>
          <a class="dropdown-item" href="https://www.genecards.org/cgi-bin/carddisp.pl?gene={{ symbol }}" target="_blank">
            Gene @GeneCards
          </a>
          <a class="dropdown-item" href="{% if entry.entrez_id %}https://www.ncbi.nlm.nih.gov/gene/{{ entry.entrez_id }}{% else %}https://www.ncbi.nlm.nih.gov/gene/?term=({{ symbol }}+AND+&quot;Homo+sapiens&quot;){% endif %}" target="_blank">
            Gene @Entrez
          </a>
          <a class="dropdown-item" href="https://www.genenames.org/cgi-bin/gene_symbol_report?match={{ symbol }}" target="_blank">
            Gene @HGNC
          </a>
          <a class="dropdown-item" href="http://www.hgmd.cf.ac.uk/ac/gene.php?gene={{ symbol }}" target="_blank">
            Gene @HGMD Public
          </a>
          <a class="dropdown-item" href="https://www.proteinatlas.org/search/{{ symbol }}" target="_blank">
            Gene @ProteinAtlas
          </a>
          <a class="dropdown-item" href="https://www.ncbi.nlm.nih.gov/pubmed/?term={{ symbol }}" target="_blank">
            Gene @PubMed
          </a>
          <a class="dropdown-item" href="https://www.ncbi.nlm.nih.gov/clinvar/?term={{ symbol }}" target="_blank">
            Gene @ClinVar
          </a>
          <a class="dropdown-item" href="https://{% if entry.release == "GRCh37" %}grch37{% else %}www{% endif %}.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g={{ entry.ensembl_gene_id }}" target="_blank">
            Gene @EnsEMBL
          </a>
          <a class="dropdown-item" href="https://stuart.radboudumc.nl/metadome/dashboard/transcript/{{ entry.ensembl_transcript_id }}/" target="_blank">
            Gene @MetaDome
          </a>
          <a class="dropdown-item" href="https://panelapp.genomicsengland.co.uk/panels/entities/{{ symbol }}" target="_blank">
            Gene @PanelApp
          </a>
          {% if hgmd_pro_enabled %}
            <a class="dropdown-item" href="{{ hgmd_pro_prefix }}/gene.php?gene={{ symbol }}" target="_blank">
              Gene @HGMD Professional
            </a>
          {% endif %}
          <a class="dropdown-item" href="https://gnomad.broadinstitute.org/gene/{{ entry.ensembl_gene_id }}" target="_blank">
            Gene @gnomAD
          </a>
          {% if entry.mgi_id %}
          <a class="dropdown-item" href="http://www.informatics.jax.org/marker/{{ entry.mgi_id }}" target="_blank">
            Gene @MGI
          </a>
          {% else %}
          <span class="dropdown-item disabled">
            Gene @MGI
          </span>
          {% endif %}
          {% if entry.hgnc_id %}
          <a class="dropdown-item" href="https://search.thegencc.org/genes/{{ entry.hgnc_id }}" target="_blank">
            Gene @GenCC
          </a>
          {% else %}
          <span class="dropdown-item disabled">
            Gene @GenCC
          </span>
          {% endif %}
          {% if entry.uniprot_ids %}
          <a class="dropdown-item" href="http://missense3d.bc.ic.ac.uk:8080/search_direct?uniprot={{ entry.uniprot_ids }}" target="_blank">
            Gene @Missense3D-DB
          </a>
          {% else %}
          <span class="dropdown-item disabled">
            Gene @Missense3D-DB
          </span>
          {% endif %}
          {% if hpoterms %}
            <a class="dropdown-item" href="https://www.ncbi.nlm.nih.gov/pubmed/?term={{ entry|get_pubmed_linkout:hpoterms }}" target="_blank">
              Gene+Phenotypes @PubMed
            </a>
          {% else %}
            <span class="dropdown-item disabled" data-toggle="tooltip" title="Enter HPO terms at More &gt; Prioritization">
              Gene+Phenotypes @PubMed
            </span>
          {% endif %}
        </div>
      </div>
    {% endif %}
    {% if entry.acmg_symbol %}
      <i class="iconify pl-2 text-danger" data-icon="fa-solid:user-md" data-toggle="tooltip" data-placement="top" title="Gene in ACMG incidental findings gene list."></i>
    {% else %}
      <i class="iconify pl-2 text-muted" data-icon="fa-solid:user-md" style="opacity: 0.2"></i>
    {% endif %}
    &nbsp;
    {% if entry.modes_of_inheritance %}
      {% for mode_of_inheritance in entry.modes_of_inheritance|listsort %}
        <span class="badge badge-info" data-toggle="tooltip" title="{{ mode_of_inheritance|mode_of_inheritance_description }}">{{ mode_of_inheritance }}</span>
      {% endfor %}
    {% endif %}
  </td>
  <td data-order="{% if entry.disease_gene %}1{% else %}0{% endif %}"><i class="iconify {% if entry.disease_gene %}text-danger"{% else %}text-muted" style="opacity: 20%"{% endif %} data-icon="fa-regular:lightbulb"></i></td>
  <td class="text-nowrap effect-change">
    <span data-toggle="tooltip" title="{{ entry.effect|join:', ' }}">
      {% with entry.hgvs_p|default:"-" as hgvs_p %}
        {% if hgvs_p == "-" or hgvs_p == "p.?" or hgvs_p == "p.=" %}
          {{ entry.hgvs_c|default:"-"|truncatechars:15 }}
        {% else %}
          {{ entry.hgvs_p|default:"-"|truncatechars:15 }}
        {% endif %}
      {% endwith %}
    </span>
    {% if entry.effect_ambiguity %}
      <i class="iconify text-muted" data-icon="fa-solid:info-circle" data-toggle="tooltip" title="The impact predictions for RefSeq and ENSEMBL transcripts differ."></i>
    {% endif %}
  </td>
  <td class="text-nowrap effect-text">{{ entry.effect|join:', ' }}</td>
  <td class="text-nowrap effect-change-protein">{{ entry.hgvs_p|default:'-' }}</td>
  <td class="text-nowrap effect-change-cdna">{{ entry.hgvs_c|default:'-' }}</td>
  <td class="text-nowrap distance-to-splicesite">{{ entry.exon_dist|default:'-' }}</td>
  {% if entry.extra_annos is not None %}
    {% for value in entry.extra_annos.0 %}
      <td class="text-nowrap extra_annos-{{ forloop.counter }}">{{ value|default:'-' }}</td>
    {% endfor %}
  {% else %}
    {% for value in result_extra_annos_header %}
      <td class="text-nowrap extra_annos-{{ forloop.counter }}">{{ '-' }}</td>
    {% endfor %}
  {% endif %}
  {% if query_type == "project" %}
    <td>{{ entry.affected_cases_per_gene|default:0 }}</td>
  {% endif %}
  {% if has_phenotype_scores %}
    <td class="text-right">
      {{ entry.phenotype_score|floatformat:3 }}
    </td>
    <td class="text-muted text-right" data-order="{{ entry.phenotype_rank }}">
      #{{ entry.phenotype_rank }}
    </td>
  {% endif %}
  {% if has_pathogenicity_scores %}
    <td class="text-right text-nowrap" data-order="{{ entry.pathogenicity_score|default:0.0 }}">
      {% if patho_score == "mutationtaster" %}
        <span data-toggle="tooltip" data-placement="top" data-html="true" title="Prediction: {{ entry.pathogenicity_score_info.prediction|default:"-" }}<br/>Model: {{ entry.pathogenicity_score_api_result.model|default:"-" }}">
          {% if not entry.pathogenicity_score_info or entry.pathogenicity_score_info.note == "error" %}
            -1 / -1.000
          {% else %}
            {{ entry.pathogenicity_score_info|mutationtaster_rank }}
            /
            {{ entry.pathogenicity_score_info.bayes_prob_dc|default:-1|mutationtaster_scale_bayes|floatformat:3 }}
          {% endif %}
        </span>
      {% else %}
        {{ entry.pathogenicity_score|floatformat:1|default:0.0 }}
      {% endif %}
    </td>
    <td class="text-muted text-right" data-order="{{ entry.pathogenicity_rank }}">
      #{{ entry.pathogenicity_rank }}
    </td>
  {% endif %}
  {% if has_phenotype_scores and has_pathogenicity_scores %}
    <td class="text-right">
      {{ entry.joint_score|floatformat:1 }}
    </td>
    <td class="text-muted text-right" data-order="{{ entry.join_rank }}">
      #{{ entry.joint_rank }}
    </td>
  {% endif %}
  {% if query_type == "case" %}
    {# pedigree members #}
    {% for member in pedigree %}
      {% include "variants/filter_result/_gt_field.html" with name=member.patient %}
    {% endfor %}
  {% else %}  {# query_type == "project" #}
    <td style="max-width: 120px;">
      <div class="sodar-overflow-container sodar-overflow-hover">
        {{ entry.family_name }}
      </div>
    </td>
    <td style="max-width: 120px;">
      <div class="sodar-overflow-container sodar-overflow-hover">
        {{ entry.sample_name|only_source_name }}
      </div>
    </td>
    {% include "variants/filter_result/_gt_field.html" with name=entry.sample_name %}
  {% endif %}
  <td>
    <div class="btn-group sodar-list-btn-group pr-2">
      <div class="btn-group sodar-list-btn-group">
        <button
          class="btn btn-primary sodar-list-btn"
          type="button"
          data-toggle="modal"
          data-target="#variantcarriermodal-{{ entry.release }}-{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}"
          onclick="showVariantCarrierModal(this, '{{ entry.release }}', '{{ entry.chromosome }}', '{{ entry.start }}', '{{ entry.reference }}', '{{ entry.alternative }}');"
        >
          <span class="iconify" data-icon="mdi:binoculars"></span>
        </button>
        {% if entry.release == "GRCh37" %}
          <a class="btn btn-primary dropdown-toggle sodar-list-dropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
            MT
          </a>
          <div class="dropdown-menu dropdown-menu-right">
            <a
              href="https://www.genecascade.org/MTc2021/ChrPos102.cgi?chromosome={{ entry.chromosome }}&position={{ entry.start }}&ref={{ entry.reference }}&alt={{ entry.alternative }}"
              data-toggle="tooltip"
              target="_blank"
              title="Open variant in MutationTaster 2021"
              class="btn btn-primary dropdown-item">
              MT 2021
            </a>
            <a
              href="https://www.genecascade.org/MT85/ChrPos85.cgi?chromosome={{ entry.chromosome }}&position={{ entry.start }}&ref={{ entry.reference }}&alt={{ entry.alternative }}"
              data-toggle="tooltip"
              target="_blank"
              title="Open variant in MutationTaster 85"
              class="btn btn-primary dropdown-item">
              MT 85
            </a>
          </div>
        {% else %}
          <span
            class="btn btn-disabled sodar-list-btn btn-outline-secondary"
            data-toggle="tooltip"
            title="MutationTaster not available for genome {{ entry.release }}"
          >
            MT
          </span>
        {% endif %}
      </div>
      <div class="btn-group sodar-list-btn-group">
        {% if ga4gh_beacon_network_widget_enabled or kiosk_mode %}
          <a
             data-toggle="popover"
             tabindex="0"
             data-trigger="focus"
             data-content="<iframe
    src='https://beacon-network.org:443/#/widget?rs={{ entry.release }}&chrom={{ entry.chromosome }}&pos={{ entry.start }}&ref={{  entry.reference }}&allele={{ entry.alternative }}'
    style='width: 100%; height: 300px; overflow: auto;'
    marginwidth='0' marginheight='0' frameborder='0' vspace='0' hspace='0'>
  </iframe>"
             data-html="true"
             data-tooltip="tooltip"
             title="Query GA4GH Beacon Network"
             class="btn btn-warning sodar-list-btn"
            >
            <i class="iconify" data-icon="fa-solid:podcast" aria-hidden="true"></i>
          </a>
        {% endif %}
        <button
          type="button"
          onclick="javascript:$.ajax({url: 'http://127.0.0.1:60151/goto?locus=chr{{ entry.chromosome }}:{{ entry.start }}-{{ entry.end }}'})"
          title="Go to locus in IGV"
          class="btn btn-secondary sodar-list-btn"
        >
          IGV
        </button>
        <button type="button" class="btn btn-secondary dropdown-toggle dropdown-toggle-split sodar-list-dropdown" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
          <span class="sr-only">Toggle Dropdown</span>
        </button>
        <div class="dropdown-menu dropdown-menu-right" style="z-index: 1030;">
          {% linkout_available_for "ucsc" entry.release as ucsc_available %}
          {% if ucsc_available %}
            <a class="dropdown-item" href="https://genome-euro.ucsc.edu/cgi-bin/hgTracks?db=hg19&position={{ entry.chromosome }}:{{ entry.start }}-{{ entry.end }}" target="_blank">
              Locus @UCSC
            </a>
          {% else %}
            <span class="dropdown-item disabled" data-toggle="tooltip" title="UCSC browser not available for reference {{ entry.release }}">
              Locus @UCSC
            </span>
          {% endif %}

          {% linkout_available_for "ensembl" entry.release as ensembl_available %}
          {% if ensembl_available %}
            <a class="dropdown-item" href="https://{% if entry.release == "GRCh37" %}grch37{% else %}www{% endif %}.ensembl.org/Homo_sapiens/Location/View?r={{ entry.chromosome }}:{{ entry.start }}-{{ entry.end }}" target="_blank">
              Locus @EnsEMBL
            </a>
          {% else %}
            <span class="dropdown-item disabled" data-toggle="tooltip" title="ENSEMBL browser not available for reference {{ entry.release }}">
              Locus @EnsEMBL
            </span>
          {% endif %}

          {% linkout_available_for "dgv" entry.release as dgv_available %}
          {% if dgv_available %}
            <a class="dropdown-item" href="http://dgv.tcag.ca/gb2/gbrowse/dgv2_hg19/?name={{ entry.chromosome }}:{{ entry.start }}-{{ entry.end }};search=Search" target="_blank">
              Locus @DGV
            </a>
          {% else %}
            <span class="dropdown-item disabled" data-toggle="tooltip" title="DGV browser not available for reference {{ entry.release }}">
              Locus @DGV
            </span>
          {% endif %}

          {% linkout_available_for "gnomad" entry.release as gnomad_available %}
          {% if gnomad_available %}
            <a class="dropdown-item" href="http://gnomad.broadinstitute.org/region/{{ entry.chromosome }}-{{ entry.start }}-{{ entry.end }}" target="_blank">
            Locus @gnomAD
            </a>
          {% else %}
            <span class="dropdown-item disabled" data-toggle="tooltip" title="gnomAD browser not available for reference {{ entry.release }}">
              Locus @gnomAD
            </span>
          {% endif %}

          {% if hgmd_pro_enabled %}
            <a class="dropdown-item" href="{{ hgmd_pro_prefix }}/gene.php?gene={{ symbol }}" target="_blank">
              Gene @HGMD Professional
            </a>
          {% endif %}

          <div class="dropdown-divider"></div>

          <a href="javascript:humanSplicingFinder('{{ symbol }}', '{{ entry.hgvs_c }}');"
            data-toggle="tooltip"
            title="Query Human Splicing Finder"
            class="dropdown-item">
            Query Human Splicing Finder
          </a>

          <a href="https://varseak.bio/ssp.php?gene={{ symbol|urlencode }}&hgvs={{ entry.hgvs_c|urlencode }}&&transcript={{ entry.refseq_transcript_id }}"
            data-toggle="tooltip"
            title="Query varSEAK Splice Site Prediction"
            target="_blank"
            class="dropdown-item">
            Query varSEAK Splicing
          </a>

          {% if "missense_variant" in entry.effect %}
            <a href="javascript:polyPhen2('{{ symbol }}', '{{ entry.hgvs_p }}');"
              data-toggle="tooltip"
              title="Query PolyPhen 2"
              class="dropdown-item">
              Query PolyPhen 2
            </a>
          {% else %}
            <span class="dropdown-item disabled">
              PolyPhen 2: missense only
            </span>
          {% endif %}

          {% linkout_available_for "umd" entry.release as umd_available %}
          {% if umd_available %}
            {% if entry.reference|length != 1 or entry.alternative|length != 1 %}
              <span class="dropdown-item disabled" data-toggle="tooltip" title="UMD Predictor can only predict SNVs">
                UMD Predictor: SNV only
              </span>
            {% elif umd_predictor_api_token %}
              <a href="{% linkout_base_url "umd" entry.release %}chromosome=chr{{ entry.chromosome }}&c_position={{ entry.start }}&wt_nucleotide={{ entry.reference }}&mutant_nucleotide={{ entry.alternative }}&token={{ umd_predictor_api_token }}"
                data-toggle="tooltip"
                title="Query UMD Predictor"
                target="_blank"
                class="dropdown-item">
                Query UMD Predictor
              </a>
              {#   #}
            {% else %}
              <span class="dropdown-item disabled" data-toggle="tooltip" title="Configure your UMD Predictor API token in the user settings.">
                UMD Predictor: no key
              </span>
            {% endif %}
          {% else %}
            <span class="dropdown-item disabled" data-toggle="tooltip" title="UMD not available for reference {{ entry.release }}">
              UMD Predictor
            </span>
          {% endif %}

          {% linkout_available_for "varsome" entry.release as varsome_available %}
          {% if varsome_available %}
            <a href="{% linkout_base_url "varsome" entry.release %}{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}"
              data-toggle="tooltip"
              title="View variant in varsome"
              target="_blank"
              class="dropdown-item">
              Query Varsome
            </a>
          {% else %}
            <span class="dropdown-item disabled" data-toggle="tooltip" title="Varsome not available for reference {{ entry.release }}">
              Query Varsome
            </span>
          {% endif %}

          {% linkout_available_for "variant_validator" entry.release as vv_available %}
          {% if vv_available %}
            <a
              data-toggle="modal"
              data-target="#variantvalidatormodal-{{ entry.release }}-{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}"
              title="Submit variant to VariantValidator"
              onclick="if ($('#variantvalidatormodal-{{ entry.release }}-{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}').find('.variant-validator-results').children().length == 0){$('#variantvalidatormodalsubmit-{{ entry.release }}-{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}').click();}"
              class="dropdown-item">
              Query VariantValidator
            </a>
          {% else %}
            <span class="dropdown-item disabled" data-toggle="tooltip" title="VariantValidator not available for reference {{ entry.release }}">
              Query VariantValidator
            </span>
          {% endif %}
        </div>
      </div>
    </div>
    <!-- Modal -->
    <div class="modal fade" id="variantcarriermodal-{{ entry.release }}-{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}">
      <div class="modal-dialog modal-lg" role="document">
        <div class="modal-content">
          <div class="modal-header">
            <h4 class="modal-title" id="variantcarriermodallabel-{{ entry.release }}-{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}">In-House Variant Carriers</h4>
            <button type="button" class="close" data-dismiss="modal" aria-label="Close">
              <span aria-hidden="true">&times;</span>
            </button>
          </div>
          <div class="modal-body">
            <div class="variant-carrier-results container-fluid">
            </div>
          </div>
          <div class="modal-footer">
            <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
          </div>
        </div>
      </div>
    </div>
    <div class="modal fade" id="variantvalidatormodal-{{ entry.release }}-{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}" tabindex="-1" role="dialog" aria-labelledby="variantvalidatormodallabel-{{ entry.release }}-{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}" aria-hidden="true">
      <div class="modal-dialog modal-lg" role="document">
        <div class="modal-content">
          <div class="modal-header">
            <h4 class="modal-title" id="variantvalidatormodallabel-{{ entry.release }}-{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}">Run through VariantValidator.org</h4>
            <button type="button" class="close" data-dismiss="modal" aria-label="Close">
              <span aria-hidden="true">&times;</span>
            </button>
          </div>
          <div class="modal-body">
            <div class="variant-validator-results container-fluid">
            </div>
          </div>
          <div class="modal-footer">
            <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
            <button class="btn btn-primary" id="variantvalidatormodalsubmit-{{ entry.release }}-{{ entry.chromosome }}-{{ entry.start }}-{{ entry.reference }}-{{ entry.alternative }}" type="button" onclick="queryVariantValidatorApi(this, '{{ entry.release }}', '{{ entry.chromosome }}', '{{ entry.start }}', '{{ entry.reference }}', '{{ entry.alternative }}');">
              <img src="/icons/fa-solid/cloud-upload-alt.svg" /> Re-Submit
            </button>
          </div>
        </div>
      </div>
    </div>
  </td>
</tr>
{% endwith %}
